### PR TITLE
Visual Editor: Keep the block at the same position while moving it

### LINKED ISSUE #161: UI: Moving block should scroll window with the block
Tall blocks, such as long paragraphs, can cause disorientation when moving blocks. 

Imagine a document:
(Small Para A)
(Large Long Para)
(Small Para B)

when moving Small Para A down and then back up the screen jumps around and it can be hard to follow what is happening. You can see this in 

It would be interesting for a designer to evaluate if a small and fast animation might help orient the user as to what is happening.

### PR DESCRIPTION
**Pull request by youknowriad**
closes #161 

Nothing too fancy here. I'm using the lifecycle methods to save the scroll position before the update and restore it after.

### CODE REVIEW

**Comment by jasmussen** [comment]
Works great! Thumbs up from a UX perspective 👍

**Review by jasmussen — APPROVED** [review]

File: editor/modes/visual-editor/block.js [hunk 1]
```diff
@@ -11,84 +11,116 @@ import Toolbar from 'components/toolbar';
 import BlockMover from 'components/block-mover';
 import BlockSwitcher from 'components/block-switcher';
 
-function VisualEditorBlock( props ) {
-	const { block } = props;
-	const settings = wp.blocks.getBlockSettings( block.blockType );
+class VisualEditorBlock extends wp.element.Component {
+	constructor() {
+		super( ...arguments );
+		this.bindBlockNode = this.bindBlockNode.bind( this );
```
Comment (aduth) [inline editor/modes/visual-editor/block.js]: Does this function need to be bound?

File: editor/modes/visual-editor/block.js [hunk 2]
```diff
@@ -11,84 +11,116 @@ import Toolbar from 'components/toolbar';
...
-		'is-selected': isSelected,
-		'is-hovered': isHovered
-	} );
+	componentDidUpdate() {
+		if ( this.previousOffset ) {
+			window.scrollTo(
+				window.scrollX,
+				window.scrollY + this.node.getBoundingClientRect().top - this.previousOffset
+			);
+			this.previousOffset = false;
```
Comment (aduth) [inline editor/modes/visual-editor/block.js]: I'm not a fan of using `false` as an empty state. Could we assign to `null`, `undefined`, or even `delete this.previousOffset` instead?

File: editor/modes/visual-editor/block.js [hunk 3]
```diff
@@ -11,84 +11,116 @@ import Toolbar from 'components/toolbar';
...
 
-	if ( ! BlockEdit ) {
-		return null;
+	componentWillReceiveProps( newProps ) {
+		if (
+			this.props.order !== newProps.order &&
+			this.props.isSelected &&
+			newProps.isSelected
+		) {
+			this.previousOffset = this.node.getBoundingClientRect().top;
```
Comment (aduth) [inline editor/modes/visual-editor/block.js]: Would it have been possible to simply store `window.scrollY` here and restore it directly in the `DidUpdate`?

File: editor/modes/visual-editor/block.js [hunk 4]
```diff
@@ -11,84 +11,116 @@ import Toolbar from 'components/toolbar';
...
-	}
 
-	function maybeDeselect( event ) {
-		// Annoyingly React does not support focusOut and we're forced to check
-		// related target to ensure it's not a child when blur fires.
-		if ( ! event.currentTarget.contains( event.relatedTarget ) ) {
-			onDeselect();
+		const { onChange, onSelect, onDeselect, onMouseEnter, onMouseLeave, onInsertAfter } = this.props;
+
+		function setAttributes( attributes ) {
```
Comment (aduth) [inline editor/modes/visual-editor/block.js]: If we're making this a proper class, these might ought to be proper member functions now.

File: editor/modes/visual-editor/block.js [hunk 5]
```diff
@@ -11,84 +11,116 @@ import Toolbar from 'components/toolbar';
...
-	);
-	/* eslint-enable jsx-a11y/no-static-element-interactions */
+		);
+		/* eslint-enable jsx-a11y/no-static-element-interactions */
+	}
 }
 
 export default connect(
 	( state, ownProps ) => ( {
+		order: state.blocks.order.findIndex( uid => ownProps.uid === uid ),
```
Comment (aduth) [inline editor/modes/visual-editor/block.js]: We haven't enabled the [polyfill](http://babeljs.io/docs/usage/polyfill/), so this will error in IE11. We might just consider enabling the polyfill, but in this case couldn't this just be:

```js
order: state.blocks.order.indexOf( ownProps.uid )
```

File: editor/modes/visual-editor/block.js (hunk 1 above)
Comment (youknowriad) [inline editor/modes/visual-editor/block.js]: yes?

File: editor/modes/visual-editor/block.js (hunk 3 above)
Comment (youknowriad) [inline editor/modes/visual-editor/block.js]: No, actually we don't want to keep the same `scrollY` since the block is moving, we want the block to show in the same position.

File: editor/modes/visual-editor/block.js (hunk 2 above)
Comment (youknowriad) [inline editor/modes/visual-editor/block.js]: Sounds good, do you have a preference for one over another? `delete` ?

Comment (aduth) [inline editor/modes/visual-editor/block.js]: >Sounds good, do you have a preference for one over another? `delete` ?

Sorta ties into a question on whether we want to be extremely clear about the existence of the instance property by initializing it either in the constructor or as an instance initializer (I don't recall if this is a spec feature or one of the staged ones).

As it is here without initialization, I'd say `delete` is most appropriate.
Otherwise, if we were to initialize, probably initialize to `null` and reset to `null` here.

Comment (youknowriad) [inline editor/modes/visual-editor/block.js]: I'll go for initialization :)

**Review by aduth — APPROVED** [review]
